abstract-spec/ffg.als

```diff
@@ -0,0 +1,208 @@
+module abs/ffg --- an abstraction of FFG
+
+open util/integer
+
+sig Payload {}
+sig Signature {}
+fact atLeastFourSignatures { #Signature >= 4 }
+
+sig Block {
+	slot: Int,
+	body: Payload,
+	parent: Block
+}
+
+one sig GenesisPayload extends Payload {}
+one sig GenesisBlock extends Block {}
+
+sig Checkpoint {
+	block: Block,
+	slot: Int
+}
+
+one sig GenesisCheckpoint extends Checkpoint {}
+
+fact genesisCheckpoint {
+	GenesisCheckpoint.block = GenesisBlock
+	GenesisCheckpoint.slot = 0
+}
+
+sig FfgVote {
+	source: Checkpoint,
+	target: Checkpoint
+}
+
+sig Vote {
+	validator: Signature,
+	ffgVote: FfgVote
+}
+
+one sig JustifiedCheckpoints {
+	justified: set Checkpoint
+}
+
+fact voteStructuralEquality {
+	all v1: Vote, v2: Vote |
+		v1 = v2 <=> v1.validator = v2.validator and v1.ffgVote = v2.ffgVote
+}
+
+fact checkpointStructuralEquality {
+	all c1: Checkpoint, c2: Checkpoint |
+		c1 = c2 <=> c1.block = c2.block and c1.slot = c2.slot
+}
+
+fact ffgVoteStructuralEquality {
+	all v1: FfgVote, v2: FfgVote |
+		v1 = v2 <=> v1.source = v2.source and v1.target = v2.target
+}
+
+fact blockStructuralEquality {
+	all b1: Block, b2: Block |
+		b1 = b2 <=> b1.slot = b2.slot and b1.body = b2.body and b1.parent = b2.parent
+}
+
+fact genesisStructure {
+	GenesisBlock.body = GenesisPayload
+	GenesisBlock.parent = GenesisBlock
+}
+
+fact slotZeroIsGenesis {
+	all b: Block | b.slot = 0 => b = GenesisBlock
+}
+
+fact noOtherBlockIsGenesis {
+	all b: Block | 
+		b.body = GenesisPayload implies b.slot = 0
+}
+
+fact blocksAreValid {
+	all b: Block | {
+		b = GenesisBlock or b.slot > b.parent.slot
+	}
+}
+
+fact ffgVotesAreValid {
+	all v: FfgVote | {
+		isAncestorDescendant[v.source.block, v.target.block]
+		v.source.slot < v.target.slot
+	}
+}
+
+fact checkpointsAreValid {
+	all c: Checkpoint | {
+		(c.slot = 0 and c.block = GenesisBlock) or c.slot > c.block.slot
+	}
```

Comment:
Question for @banhday and @saltiniroberto:

This constraint is in the paper, but not in the Python spec.
Should we keep it?